gse_industry_fsm/models/project.py

```diff
@@ -59,3 +59,38 @@ def action_fs_navigate(self):
     def onchange_partner_service_id(self):
         for task in self:
             task.partner_service_id = task.partner_id.address_get(['field_service'])['field_service'] if task.partner_id else False
+
+    @api.onchange("stage_id")
+    def onchange_stage_id(self):
+        if self.stage_id and self.stage_id.name == 'Done':
+            user = self.env.user
+            if not user.has_group('gse_access_rights.group_stock_validator'):
+                raise AccessError("You don't have permission to perform this action.")
+            
+            if not self.is_task_done():
+                raise AccessError("Mandory steps uncompleted.")
+            
+
+    def action_fsm_validate(self, stop_running_timers=False):
+        """ If All product are delivered, if the Worksheet is completed, At least one timesheet entry is recorded for the task and user has privileges :
+           Task can be mark as done.
+        """
+        user = self.env.user
+        if not user.has_group('gse_access_rights.group_stock_validator'):
+            raise AccessError("You don't have permission to perform this action. Kindly contact your administrator.")
+        
+        if not self.is_task_done():
+            raise AccessError("Mandory steps uncompleted.")
```

Comment:
@iambenkis 
Can you be more explicit in the Error Message:
- Some products still need to be delivered.
- The Timesheet aren't completed.